File: sdk-java/dephealth-core/src/main/java/biz/kryukov/dev/dephealth/Dependency.java
package biz.kryukov.dev.dephealth;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Dependency descriptor: name, type, endpoints, and check configuration. Immutable.
 */
public final class Dependency {

    private static final Pattern NAME_PATTERN = Pattern.compile("^[a-z][a-z0-9-]*$");
    private static final int MAX_NAME_LENGTH = 63;

    private final String name;
    private final DependencyType type;
    private final boolean critical;
    private final List<Endpoint> endpoints;
    private final CheckConfig config;

    private Dependency(Builder builder) {
        this.name = builder.name;
        this.type = builder.type;
        this.critical = builder.criticalValue;
        this.endpoints = Collections.unmodifiableList(builder.endpoints);
        this.config = builder.config;
    }

    /** Returns the dependency name. */
    public String name() {
        return name;
    }

    /** Returns the dependency type. */
    public DependencyType type() {
        return type;
    }

    /** Returns whether this dependency is marked as critical. */
    public boolean critical() {
        return critical;
    }

    /** Returns the immutable list of endpoints. */
    public List<Endpoint> endpoints() {
        return endpoints;
    }

    /** Returns the health check configuration. */
    public CheckConfig config() {
        return config;
    }

    /**
     * Converts a boolean to "yes"/"no" string for the critical label.
     */
    public static String boolToYesNo(boolean value) {
        return value ? "yes" : "no";
    }

    /**
     * Validates that the dependency name matches the naming rules.
     *
     * @param name dependency name
     * @throws ValidationException if the name is invalid
     */
    public static void validateName(String name) {
        Objects.requireNonNull(name, "dependency name");
        if (name.isEmpty() || name.length() > MAX_NAME_LENGTH) {
            throw new ValidationException(
                    "dependency name must be 1-" + MAX_NAME_LENGTH + " characters, got '"
                            + name + "' (" + name.length() + " chars)");
        }
        if (!NAME_PATTERN.matcher(name).matches()) {
            throw new ValidationException(
                    "dependency name must match " + NAME_PATTERN.pattern()
                            + ", got '" + name + "'");
        }
    }

    /**
     * Creates a new builder for a dependency.
     *
     * @param name dependency name
     * @param type dependency type
     * @return a new builder instance
     */
    public static Builder builder(String name, DependencyType type) {
        return new Builder(name, type);
    }

    /** Builder for {@link Dependency}. */
    public static final class Builder {
        private final String name;
        private final DependencyType type;
        private Boolean criticalValue;
        private List<Endpoint> endpoints = List.of();
        private CheckConfig config = CheckConfig.defaults();

        private Builder(String name, DependencyType type) {
            this.name = Objects.requireNonNull(name, "name");
            this.type = Objects.requireNonNull(type, "type");
        }

        /**
         * Sets the criticality of the dependency. Required parameter.
         */
        public Builder critical(boolean critical) {
            this.criticalValue = critical;
            return this;
        }

        /** Sets the list of endpoints. */
        public Builder endpoints(List<Endpoint> endpoints) {
            this.endpoints = List.copyOf(endpoints);
            return this;
        }

        /** Sets a single endpoint. */
        public Builder endpoint(Endpoint endpoint) {
            this.endpoints = List.of(endpoint);
            return this;
        }

        /** Sets the health check configuration. */
        public Builder config(CheckConfig config) {
            this.config = Objects.requireNonNull(config, "config");
            return this;
        }

        /** Builds and validates the dependency. */
        public Dependency build() {
            validate();
            return new Dependency(this);
        }

        private void validate() {
            validateName(name);
            if (endpoints.isEmpty()) {
                throw new ValidationException("dependency '" + name + "' must have at least one endpoint");
            }
            if (criticalValue == null) {
                throw new ValidationException(
                        "dependency '" + name + "' must have critical flag set explicitly");
            }
            // Validate labels of each endpoint
            for (Endpoint ep : endpoints) {
                Endpoint.validateLabels(ep.labels());
            }
        }
    }
}

File: sdk-csharp/DepHealth.Core/Dependency.cs
using System.Collections.ObjectModel;
using System.Text.RegularExpressions;

namespace DepHealth;

/// <summary>
/// Dependency descriptor: name, type, set of endpoints, check configuration. Immutable.
/// </summary>
public sealed partial class Dependency
{
    private const int MaxNameLength = 63;

    [GeneratedRegex("^[a-z][a-z0-9-]*$")]
    private static partial Regex NamePattern();

    /// <summary>Dependency name (lowercase, alphanumeric with hyphens).</summary>
    public string Name { get; }

    /// <summary>Type of the dependency (e.g. Http, Postgres, Redis).</summary>
    public DependencyType Type { get; }

    /// <summary>Whether this dependency is critical for the application.</summary>
    public bool Critical { get; }

    /// <summary>Endpoints to monitor for this dependency.</summary>
    public IReadOnlyList<Endpoint> Endpoints { get; }

    /// <summary>Health check configuration (intervals, timeouts, thresholds).</summary>
    public CheckConfig Config { get; }

    private Dependency(Builder builder)
    {
        Name = builder.NameValue;
        Type = builder.TypeValue;
        Critical = builder.CriticalValue!.Value;
        Endpoints = new ReadOnlyCollection<Endpoint>(new List<Endpoint>(builder.EndpointsValue));
        Config = builder.ConfigValue;
    }

    /// <summary>Converts a boolean to "yes"/"no" string.</summary>
    public static string BoolToYesNo(bool value) => value ? "yes" : "no";

    /// <summary>Creates a new builder for constructing a <see cref="Dependency"/>.</summary>
    /// <param name="name">Dependency name (lowercase, alphanumeric with hyphens, max 63 chars).</param>
    /// <param name="type">Type of the dependency.</param>
    public static Builder CreateBuilder(string name, DependencyType type) => new(name, type);

    /// <summary>
    /// Fluent builder for constructing a <see cref="Dependency"/> instance.
    /// </summary>
    public sealed class Builder
    {
        internal string NameValue;
        internal DependencyType TypeValue;
        internal bool? CriticalValue;
        internal List<Endpoint> EndpointsValue = [];
        internal CheckConfig ConfigValue = CheckConfig.Defaults();

        internal Builder(string name, DependencyType type)
        {
            NameValue = name ?? throw new ArgumentNullException(nameof(name));
            TypeValue = type;
        }

        /// <summary>Sets whether this dependency is critical.</summary>
        /// <param name="critical"><c>true</c> if the dependency is critical for the application.</param>
        public Builder WithCritical(bool critical)
        {
            CriticalValue = critical;
            return this;
        }

        /// <summary>Sets the endpoints to monitor.</summary>
        /// <param name="endpoints">Collection of endpoints.</param>
        public Builder WithEndpoints(IEnumerable<Endpoint> endpoints)
        {
            EndpointsValue = new List<Endpoint>(endpoints);
            return this;
        }

        /// <summary>Sets a single endpoint to monitor.</summary>
        /// <param name="endpoint">The endpoint.</param>
        public Builder WithEndpoint(Endpoint endpoint)
        {
            EndpointsValue = [endpoint];
            return this;
        }

        /// <summary>Sets the health check configuration.</summary>
        /// <param name="config">Check configuration with intervals, timeouts, and thresholds.</param>
        public Builder WithConfig(CheckConfig config)
        {
            ConfigValue = config ?? throw new ArgumentNullException(nameof(config));
            return this;
        }

        /// <summary>Validates and builds the <see cref="Dependency"/> instance.</summary>
        /// <exception cref="ValidationException">Thrown when validation fails.</exception>
        public Dependency Build()
        {
            Validate();
            return new Dependency(this);
        }

        private void Validate()
        {
            if (string.IsNullOrEmpty(NameValue) || NameValue.Length > MaxNameLength)
            {
                throw new ValidationException(
                    $"dependency name must be 1-{MaxNameLength} characters, got '{NameValue}' ({NameValue?.Length ?? 0} chars)");
            }

            if (!NamePattern().IsMatch(NameValue))
            {
                throw new ValidationException(
                    $"dependency name must match ^[a-z][a-z0-9-]*$, got '{NameValue}'");
            }

            if (CriticalValue is null)
            {
                throw new ValidationException(
                    $"dependency '{NameValue}': critical must be explicitly set");
            }

            if (EndpointsValue.Count == 0)
            {
                throw new ValidationException(
                    $"dependency '{NameValue}' must have at least one endpoint");
            }

            foreach (var ep in EndpointsValue)
            {
                Endpoint.ValidateLabels(ep.Labels);
            }
        }
    }
}
